File: BioVault-main/mobile-app/cpp/src/proof_of_reality.cpp
#include "proof_of_reality.h"
#include <algorithm>
#include <cmath>
#include <numeric>
#include <sstream>
#include <iomanip>

namespace biovault {
namespace reality {

double ProofOfRealityAnalyzer::calculatePearsonCorrelation(
    const std::vector<double>& signal1,
    const std::vector<double>& signal2) {
    
    if (signal1.size() != signal2.size() || signal1.empty()) {
        return 0.0;
    }

    const size_t n = signal1.size();
    
    // Calculate means
    double mean1 = std::accumulate(signal1.begin(), signal1.end(), 0.0) / n;
    double mean2 = std::accumulate(signal2.begin(), signal2.end(), 0.0) / n;
    
    // Calculate covariance and standard deviations
    double covariance = 0.0;
    double variance1 = 0.0;
    double variance2 = 0.0;
    
    for (size_t i = 0; i < n; ++i) {
        double diff1 = signal1[i] - mean1;
        double diff2 = signal2[i] - mean2;
        covariance += diff1 * diff2;
        variance1 += diff1 * diff1;
        variance2 += diff2 * diff2;
    }
    
    // Calculate Pearson correlation coefficient
    double denominator = std::sqrt(variance1 * variance2);
    if (denominator < 1e-10) {
        return 0.0;
    }
    
    return covariance / denominator;
}

std::vector<CorrelationPair> ProofOfRealityAnalyzer::analyzePulseUniqueness(
    const std::vector<PulseData>& pulseData,
    double threshold) {
    
    std::vector<CorrelationPair> results;
    
    // Calculate pairwise correlations
    for (size_t i = 0; i < pulseData.size(); ++i) {
        for (size_t j = i + 1; j < pulseData.size(); ++j) {
            const auto& pulse1 = pulseData[i];
            const auto& pulse2 = pulseData[j];
            
            if (pulse1.rawSignal.empty() || pulse2.rawSignal.empty()) {
                continue;
            }
            
            // Align signals to same length
            size_t minLen = std::min(pulse1.rawSignal.size(), pulse2.rawSignal.size());
            std::vector<double> sig1(pulse1.rawSignal.begin(), pulse1.rawSignal.begin() + minLen);
            std::vector<double> sig2(pulse2.rawSignal.begin(), pulse2.rawSignal.begin() + minLen);
            
            // Calculate correlation
            double corr = calculatePearsonCorrelation(sig1, sig2);
            
            CorrelationPair cp;
            cp.faceId1 = pulse1.faceId;
            cp.faceId2 = pulse2.faceId;
            cp.pairId = std::to_string(pulse1.faceId) + std::to_string(pulse2.faceId);
            cp.coefficient = corr;
            cp.replayAttack = (std::abs(corr) > threshold);
            
            results.push_back(cp);
        }
    }
    
    return results;
}

ProofOfRealityMetadata ProofOfRealityAnalyzer::createMetadata(
    const std::vector<PulseData>& pulseData,
    const std::string& consensusHash,
    const std::string& hardwareDNA,
    const std::string& videoFrameHash,
    uint64_t timestamp,
    const std::string& verificationStatus) {
    
    // Analyze pulse uniqueness
    auto correlations = analyzePulseUniqueness(pulseData, 0.95);
    
    return ProofOfRealityMetadata::fromAnalysis(
        pulseData,
        correlations,
        consensusHash,
        hardwareDNA,
        videoFrameHash,
        timestamp,
        verificationStatus
    );
}

ProofOfRealityMetadata ProofOfRealityMetadata::fromAnalysis(
    const std::vector<PulseData>& pulses,
    const std::vector<CorrelationPair>& correlations,
    const std::string& consensusHash_,
    const std::string& hardwareDNA_,
    const std::string& videoFrameHash_,
    uint64_t timestamp_,
    const std::string& verificationStatus_) {
    
    ProofOfRealityMetadata meta;
    meta.pulseData = pulses;
    meta.consensusHash = consensusHash_;
    meta.hardwareDNA = hardwareDNA_;
    meta.videoFrameHash = videoFrameHash_;
    meta.timestamp = timestamp_;
    meta.verificationStatus = verificationStatus_;
    meta.detectedFaces = static_cast<int>(pulses.size());
    meta.receivedSignatures = static_cast<int>(pulses.size());
    
    // Build correlation maps
    bool anyReplayAttack = false;
    for (const auto& corr : correlations) {
        meta.correlationCoefficients[corr.pairId] = corr.coefficient;
        meta.replayAttackFlags[corr.pairId] = corr.replayAttack;
        if (corr.replayAttack) {
            anyReplayAttack = true;
        }
    }
    
    meta.allUniqueSignals = !anyReplayAttack;
    
    return meta;
}

std::string ProofOfRealityMetadata::toJSON() const {
    std::ostringstream json;
    json << std::fixed << std::setprecision(4);
    
    json << "{\n";
    
    // Pulse data array
    json << "  \"pulse_data\": [\n";
    for (size_t i = 0; i < pulseData.size(); ++i) {
        const auto& pulse = pulseData[i];
        json << "    {\n";
        json << "      \"face_id\": " << pulse.faceId << ",\n";
        json << "      \"bpm\": " << pulse.bpm << ",\n";
        json << "      \"confidence\": " << pulse.confidence << ",\n";
        json << "      \"signal_length\": " << pulse.rawSignal.size() << "\n";
        json << "    }";
        if (i < pulseData.size() - 1) json << ",";
        json << "\n";
    }
    json << "  ],\n";
    
    // Correlation coefficients
    json << "  \"correlation_coefficients\": {\n";
    size_t idx = 0;
    for (const auto& pair : correlationCoefficients) {
        json << "    \"" << pair.first << "\": " << pair.second;
        if (++idx < correlationCoefficients.size()) json << ",";
        json << "\n";
    }
    json << "  },\n";
    
    // Replay attack flags
    json << "  \"replay_attack_flags\": {\n";
    idx = 0;
    for (const auto& pair : replayAttackFlags) {
        json << "    \"" << pair.first << "\": " << (pair.second ? "true" : "false");
        if (++idx < replayAttackFlags.size()) json << ",";
        json << "\n";
    }
    json << "  },\n";
    
    // Consensus data
    json << "  \"consensus_hash\": \"" << consensusHash << "\",\n";
    json << "  \"hardware_dna\": \"" << hardwareDNA << "\",\n";
    json << "  \"video_frame_hash\": \"" << videoFrameHash << "\",\n";
    json << "  \"timestamp\": " << timestamp << ",\n";
    json << "  \"verification_status\": \"" << verificationStatus << "\",\n";
    json << "  \"all_unique_signals\": " << (allUniqueSignals ? "true" : "false") << ",\n";
    json << "  \"detected_faces\": " << detectedFaces << ",\n";
    json << "  \"received_signatures\": " << receivedSignatures << "\n";
    
    json << "}";
    
    return json.str();
}

} // namespace reality
} // namespace biovault

File: BioVault-main/scripts/proof_of_reality.py
"""
Proof of Reality: Cross-Correlation Analysis for Replay Attack Detection

Python reference implementation for testing and validation.
Matches C++ implementation in proof_of_reality.cpp

Usage:
    python proof_of_reality.py <pulse1.csv> <pulse2.csv> <pulse3.csv>
"""

import numpy as np
import json
import sys
from typing import List, Dict, Tuple

def calculate_cross_correlation(pulse1: np.ndarray, pulse2: np.ndarray) -> float:
    """
    Calculate the Pearson correlation coefficient between two pulses.
    
    Parameters:
    pulse1 (numpy array): The first pulse signal.
    pulse2 (numpy array): The second pulse signal.
    
    Returns:
    float: The correlation coefficient [-1, 1].
    """
    # Handle different lengths by taking shorter
    min_len = min(len(pulse1), len(pulse2))
    pulse1 = pulse1[:min_len]
    pulse2 = pulse2[:min_len]
    
    # Pearson correlation
    return np.corrcoef(pulse1, pulse2)[0, 1]

def analyze_pulse_uniqueness(pulse_data: List[Dict], threshold: float = 0.95) -> Dict:
    """
    Analyze multiple pulse signals for uniqueness.
    
    Parameters:
    pulse_data: List of dicts with keys: face_id, bpm, signal
    threshold: Correlation threshold for replay detection (default 0.95)
    
    Returns:
    dict: Analysis results with correlations and replay flags
    """
    n_pulses = len(pulse_data)
    correlations = {}
    replay_flags = {}
    
    # Calculate pairwise correlations
    for i in range(n_pulses):
        for j in range(i + 1, n_pulses):
            pulse1 = np.array(pulse_data[i]['signal'])
            pulse2 = np.array(pulse_data[j]['signal'])
            
            face_id1 = pulse_data[i]['face_id']
            face_id2 = pulse_data[j]['face_id']
            pair_id = f"{face_id1}{face_id2}"
            
            corr = calculate_cross_correlation(pulse1, pulse2)
            correlations[pair_id] = float(corr)
            
            # Flag as replay attack if correlation > threshold
            is_replay = abs(corr) > threshold
            replay_flags[pair_id] = is_replay
            
            status = "⚠️  REPLAY ATTACK" if is_replay else "✅ Unique"
            print(f"Correlation between pulse {face_id1} and pulse {face_id2}: {corr:.4f} - {status}")
    
    # Check if all signals are unique
    all_unique = not any(replay_flags.values())
    
    return {
        'correlation_coefficients': correlations,
        'replay_attack_flags': replay_flags,
        'all_unique_signals': all_unique
    }

def create_proof_of_reality_metadata(
    pulse_data: List[Dict],
    consensus_hash: str,
    hardware_dna: str,
    video_frame_hash: str,
    timestamp: int,
    verification_status: str
) -> Dict:
    """
    Create complete Proof of Reality metadata structure.
    """
    # Analyze correlations
    analysis = analyze_pulse_uniqueness(pulse_data)
    
    metadata = {
        'pulse_data': [
            {
                'face_id': p['face_id'],
                'bpm': p['bpm'],
                'confidence': p.get('confidence', 0.0),
                'signal_length': len(p['signal'])
            }
            for p in pulse_data
        ],
        'correlation_coefficients': analysis['correlation_coefficients'],
        'replay_attack_flags': analysis['replay_attack_flags'],
        'consensus_hash': consensus_hash,
        'hardware_dna': hardware_dna,
        'video_frame_hash': video_frame_hash,
        'timestamp': timestamp,
        'verification_status': verification_status,
        'all_unique_signals': analysis['all_unique_signals'],
        'detected_faces': len(pulse_data),
        'received_signatures': len(pulse_data)
    }
    
    return metadata

def main():
    """Example usage"""
    
    # Example: 3 pulse signals
    # In production, these come from rPPG extraction
    
    if len(sys.argv) > 1:
        # Load from CSV files
        pulse_data = []
        for i, filepath in enumerate(sys.argv[1:], start=1):
            signal = np.loadtxt(filepath, delimiter=',')
            bpm = int(np.mean(signal) * 60)  # Simplified BPM estimation
            pulse_data.append({
                'face_id': i,
                'bpm': bpm,
                'signal': signal.tolist(),
                'confidence': 0.85
            })
    else:
        # Demo with synthetic signals
        print("📊 Proof of Reality Demo - Synthetic Signals\n")
        
        # Pulse 1: Normal signal (68 BPM)
        pulse1 = np.sin(2 * np.pi * 1.13 * np.arange(300) / 30) + np.random.normal(0, 0.1, 300)
        
        # Pulse 2: Nearly identical to pulse1 (REPLAY ATTACK)
        pulse2 = pulse1 + np.random.normal(0, 0.05, 300)
        
        # Pulse 3: Different signal (75 BPM)
        pulse3 = np.sin(2 * np.pi * 1.25 * np.arange(300) / 30) + np.random.normal(0, 0.1, 300)
        
        pulse_data = [
            {'face_id': 1, 'bpm': 68, 'signal': pulse1.tolist(), 'confidence': 0.85},
            {'face_id': 2, 'bpm': 68, 'signal': pulse2.tolist(), 'confidence': 0.90},
            {'face_id': 3, 'bpm': 75, 'signal': pulse3.tolist(), 'confidence': 0.88}
        ]
    
    # Create metadata
    metadata = create_proof_of_reality_metadata(
        pulse_data=pulse_data,
        consensus_hash="a3f8e9d1c2b5a7f6e4d3c2b1a9f8e7d6c5b4a3f2e1d0",
        hardware_dna="prnu_fingerprint_device_12345",
        video_frame_hash="b2e7c1d3a4f5b6e7d8c9a0b1c2d3e4f5a6b7c8d9e0f1",
        timestamp=1738454400,
        verification_status="COMPLETE"
    )
    
    # Print summary
    print(f"\n📊 Proof of Reality Summary:")
    print(f"   Detected Faces: {metadata['detected_faces']}")
    print(f"   Unique Signals: {'✅ Yes' if metadata['all_unique_signals'] else '⚠️  No (replay attack detected)'}")
    print(f"   Verification: {metadata['verification_status']}")
    
    # List replay attacks
    replay_pairs = [pair for pair, flag in metadata['replay_attack_flags'].items() if flag]
    if replay_pairs:
        print(f"\n⚠️  Replay Attacks Detected:")
        for pair in replay_pairs:
            corr = metadata['correlation_coefficients'][pair]
            print(f"   Face pair {pair}: correlation = {corr:.4f} (> 0.95)")
    
    # Save to JSON
    output_file = 'proof_of_reality.json'
    with open(output_file, 'w') as f:
        json.dump(metadata, f, indent=2)
    
    print(f"\n✅ Metadata saved to {output_file}")
    print(f"   Upload to IPFS and anchor on Polygon:")
    print(f"   node scripts/anchorProofOfReality.js {output_file} <videoIPFSCID>")

if __name__ == '__main__':
    main()
